### PR TITLE
added version 4 transformer

### LINKED ISSUE #11: WhatsApp: Create 4.0 Version

### PR DESCRIPTION
**Pull request by AlexTeusz**
Closes #11

### CODE REVIEW

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+
+	// create list for whatsapp content
+	let whatsAppContents: TWhatsAppContent[] = [];
+
+	// Loop through all provided Cogngiy bot messages
+	if (processedOutput.outputStack) {
+		for (let stackItem of processedOutput.outputStack) {
+
+			// check if default text was sent
+			if (stackItem.text && !(stackItem.data && stackItem.data?._cognigy && (stackItem.data?._cognigy?._default))) {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:151]: This if statement can be boiled down to (stackItem && !stackItem.data?._cognigy?._default) if I am not mistaken

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+						longitude,
+						latitude,
+						name,
+						address
+					},
+				});
+			}
+
+			// check if default templates are defined
+			else if (stackItem.data && (stackItem.data?._cognigy?._default)) {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:179]: can be boiled down to (stackItem.data?._cognigy?._default)

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+
+				let defaultContent = stackItem.data._cognigy._default;
+
+				// look for gallery
+				if (defaultContent._gallery != null && defaultContent._gallery?.items) {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:184]: can be boiled down to (defaultContent._gallery?.items)

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+							media: {
+								type: "image",
+								url: element.imageUrl,
+								caption: `*${element.title}*\n\n${element.subtitle}`
+							}
+						});
+
+						console.log(element)
+						// check for buttons and show them as quick replies
+						if (element.buttons && element.buttons?.length !== 0) {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:208]: can be shortened to (element.buttons?.length)

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 1 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:151]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 2 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:179]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 3 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:184]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+					// create quick replies message as message bubble
+					whatsAppContents.push({
+						from: sessionId,
+						contentType: "text",
+						text: createWhatsAppQuickReplies(quickReplies, sessionStorage)
+					});
+				}
+
+				// check for image
+				else if (defaultContent._image !== null && defaultContent._image?.type === "image") {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:242]: can be shortened

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+						media: {
+							type: "image",
+							url: defaultContent._image.imageUrl,
+							caption: defaultContent._image.fallbackText || ""
+						}
+					});
+				}
+
+				// check for audio
+				else if (defaultContent._audio !== null && defaultContent._audio?.type === "audio") {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:255]: can be shortened

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+						media: {
+							type: "audio",
+							url: defaultContent._audio.audioUrl,
+							caption: defaultContent._audio.fallbackText || ""
+						}
+					});
+				}
+
+				// check for video
+				else if (defaultContent._video !== null && defaultContent._video?.type === "video") {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:268]: can be shortened

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 4 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:208]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 5 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:242]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 6 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:255]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 7 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:268]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 8]
```diff
@@ -148,7 +148,7 @@ const convertWebchatContentToWhatsApp = (processedOutput, sessionId: string, ses
 		for (let stackItem of processedOutput.outputStack) {
 
 			// check if default text was sent
-			if (stackItem.text && !(stackItem.data && stackItem.data?._cognigy && (stackItem.data?._cognigy?._default))) {
+			if (stackItem.text && !(stackItem && !stackItem.data?._cognigy?._default)) {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:151]: simply (stackItem.text && !stackItem.data?._cognigy?._default) should do

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts [hunk 9]
```diff
@@ -203,9 +203,8 @@ const convertWebchatContentToWhatsApp = (processedOutput, sessionId: string, ses
 							}
 						});
 
-						console.log(element)
 						// check for buttons and show them as quick replies
-						if (element.buttons && element.buttons?.length !== 0) {
+						if (element.buttons?.length !== 0) {
```
Comment (cabullmann) [inline transformers/tyntecWhatsAppTransformer/transformer-v4.ts:207]: (element.buttons?.length) would be enough, if length is 0 then the statement is false

Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:207]: yes but the more secure version is then to check if the length is not null as well

Comment (cabullmann) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:207]: (element.buttons?.length) also is false if length is null

Comment (cabullmann) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:207]: (element.buttons?.length !== 0) actually is true if length is null, which would be a problem
(element.buttons?.length != 0) would not have that problem

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 8 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts:151]: done

File: transformers/tyntecWhatsAppTransformer/transformer-v4.ts (hunk 9 above)
Comment (AlexTeusz) [reply transformers/tyntecWhatsAppTransformer/transformer-v4.ts]: done

**Comment by AlexTeusz** [comment]
@cabullmann I tested the new version of the transformer and it works perfectly.

**Review by cabullmann — APPROVED** [review]
passt

**Pull request merged**